Crypto info

mgrobertso opened the pull request:
# Login form
- added fake data with json server
- some  validation 
<figure>
<img width="947" alt="Capture5" src="https://user-images.githubusercontent.com/109672728/188224299-b8d15c91-e4dd-4654-a07d-d4ed8e29105a.PNG"> 
</figure>

# Table
- Planning on changing the state of the buttons when a user  login
- for now it it renders the add buttons

<figure>
<img width="947" alt="Capture5" src="https://user-images.githubusercontent.com/109672728/188227926-f52db706-8080-4aee-8665-7417d805e72b.PNG">
</figure>
<figure>
<img width="947" alt="Capture5" src="https://user-images.githubusercontent.com/109672728/188227960-16232c7f-a117-4b56-bd52-a26f8ec27c2a.PNG">
</figure>

<figure>
<img width="947" alt="Capture5" src="https://user-images.githubusercontent.com/109672728/188678754-c9ecc0f2-7f24-494e-9bff-e470bd056217.PNG">
</figure>

swhook52 commented on .vscode/settings.json:
```diff
@@ -1,5 +1,16 @@
+  "eslint.alwaysShowStatus": false,
+  "eslint.format.enable": true,
+  "eslint.lintTask.enable": true,
 }
```
You should ignore vscode related files in your git ignore.

```
# IDE - VSCode
.vscode/*
!.vscode/settings.json
!.vscode/tasks.json
!.vscode/launch.json
!.vscode/extensions.json
.history/*
```

mgrobertso replied:
fixed, I untracked them since they were already added to GitHub

swhook52 commented on src/app/crypto-info/crypto-info.component.ts:
```diff
@@ -0,0 +1,29 @@
+  ) {}
+
+  ngOnInit(): void {
+    this.sub = this.cryptoService
```
you'll need to unsubscribe from `this.sub`

mgrobertso replied:
fixed, I unsubscribed in ngonDestroy

swhook52 commented on src/app/login/login.component.ts:
```diff
@@ -1,47 +1,56 @@
+  }
 
+  onClickSubmit(data: user) {
+    this.sub = this.http.get<any[]>('http://localhost:3000/posts').subscribe(
```
All of this logic should go in your auth service. This component should call into the service with the username and password and return back an Observable of a user or whatever info you need back from it. (that should not include returning an object with the password). Then your code here is just a call to the service and some logic redirect or show an error.

mgrobertso replied:
I moved the logic to my auth service and changed logic so it returned an appropriate data  type

swhook52 commented on src/app/shared/service/auth.service.ts:
```diff
@@ -0,0 +1,37 @@
+    this.isLoggedIn$ = new BehaviorSubject(isLoggedIn);
+  }
+
+  getState() {
```
You need to use TypeScript more here and provide return types for all of your methods that return something. ESLint should be warning you about these things too so you should make sure you are getting those warnings or you should try to configure that again.

swhook52 commented on src/app/shared/service/auth.service.ts:
```diff
@@ -0,0 +1,37 @@
+  public userInfo$: BehaviorSubject<user|null> = new BehaviorSubject<user|null>(null);
+
+  constructor(public http:HttpClient) {
+    const isLoggedIn = localStorage.getItem('loggedIn') === 'false';
```
You should work with a boolean here. convert local storage to a boolean that way you can work with booleans through the rest of the code.

mgrobertso replied:
converted the return to a boolean

swhook52 commented on src/app/login/login.component.ts:
```diff
@@ -1,56 +1,28 @@
-  }
-
-  ngOnDestroy(): void {
-    this.sub.unsubscribe();
```
you'll still need to unsubscribe

swhook52 commented on src/app/shared/service/auth.service.ts:
```diff
@@ -1,37 +1,96 @@
   providedIn: 'root',
 })
 export class AuthService {
   public isLoggedIn$: BehaviorSubject<boolean>;
```
As a best practice, you don't want to expose BehaviorSubjects to the calling code. The BehaviorSubject allows you to next onto the observable or completely change it. You wouldn't want the calling code to add problems to your service. Consider the BehaviorSubject your private reference to the observable. Then create a public reference by using `as Observable` on the BehaviorSubject.

swhook52 commented on src/app/shared/service/auth.service.ts:
```diff
@@ -1,37 +1,96 @@
+          alert('user not found');
+        }
+      });
+    return this.isLoggedIn$;
```
Remember that the subscribe content above doesn't necessarily run before this method returns this result. Sometimes it will, but most times subscribe will happen after you return from the method. Based on the inconsistency, your users will get varying results of if the user is logged in or not based on what the last value was in the observable.

Personally I think you're better off _not_ doing a subscribe and instead returning the observable back to the caller. Then you don't even have to manage the unsubscribe in this service. You would move all the code from the subscribe into a map if you want to return back the true or false. Or you could put all the logic into a tap and not return anything if you preferred that the client look for changes on the isLoggedIn stream.

swhook52 commented on src/app/shared/service/auth.service.ts:
```diff
@@ -1,37 +1,96 @@
+          return err;
+        })
+      );
+    return userMade;
```
Same comment here. userMade will almost always return false in a real API scenario because the post will take time but you would have already returned from the method. You need to convert this to return the whole observable instead of storing it in `this.regSub`. Then instead of the subscribe, you'll have a map that returns true or false based on what happened.

Something is also not right here because you're piping on a subscription. all pipes happen before a subscribe.

swhook52 commented on src/app/crypto-data-component/crypto-data-component.component.ts:
```diff
@@ -77,19 +77,20 @@ export class CryptoDataComponentComponent implements OnInit, OnDestroy {
   removeWatch(id: string) {
     if (this.auth.isLoggedIn$) {
-      const userValue = this.auth.userInfo$.value as user;
+      const userValue = this.auth.userInfo$.value as User;
```
I made comments in the service that will change this line. using value isn't a good practice here and after only exposing an observable and not a BehaviorSubject you won't be able to use `.value`. You'll need to change this to be using an observable so you'll likely be subscribing (and unsubscribing).

swhook52 commented on src/app/shared/service/auth.service.ts:
```diff
@@ -1,96 +1,83 @@
+
+  addWatch(id: string): void {
+    this._userInfo.pipe(
+      map((watch) => {
```
This should be a `tap` because it's a side effect. You don't plan on using the mapped value.

swhook52 commented on src/app/shared/service/auth.service.ts:
```diff
@@ -1,96 +1,83 @@
+   */
+  signup(formData: SignupRequest): Observable<User> {
+    return this.http.post<User>('http://localhost:3000/signup', formData).pipe(
+      map((user) => {
```
You don't need this map if you return the same value.

swhook52 reviewed (changes requested):
Things looking better. Good job on the `BehaviorSubject`. Let's wrap up the rest.

mgrobertso commented:
<figure>
<img width="947" alt="Capture5" src="https://user-images.githubusercontent.com/109672728/191286112-d4ed19ad-88fe-442f-995c-a25ea6671f0b.PNG">
</figure>
Fixed the footer and the changing fonts

<figure>
<img width="947" alt="Capture5" src="https://user-images.githubusercontent.com/109672728/191285754-53763886-dda4-4eb8-b785-f5eaa3be8366.PNG">
</figure>
Its a little  zoomed out but this is the detail page, the top bar isn't very reactive yet.


<figure>
<img width="947" alt="Capture5" src="https://user-images.githubusercontent.com/109672728/191287329-8bd10756-9f4a-4f3e-8343-5a3340b14d0a.PNG">
</figure>
Updated the login page a little

swhook52 commented on src/app/crypto-data-component/crypto-data-component.component.ts:
```diff
@@ -74,23 +73,28 @@ export class CryptoDataComponentComponent implements OnInit, OnDestroy {
+          this.watchList.push(id);
+          this.auth.addWatch(id);
+          console.log(
+            this.auth.userInfo$.subscribe((res) => {
```
You should just take this out because the complexity of getting it isn't worth logging it. You'd be better off logging it from your auth service if you wanted to see it.

You should not be subscribing to an observable within a subscribe. You would need to include all observables you want to access within the same observable using an operator like `combineLatest` and _many_  others.
